core/arch/arm/sm/suspend.S

```diff
@@ -0,0 +1,243 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <kernel/unwind.h>
+#include <asm.S>
+#include <asm-defines.h>
+#include <arm.h>
+#include <arm32_macros.S>
+#include <arm32_macros_cortex_a9.S>
+#include <kernel/asan.h>
+#include <platform_config.h>
+#include <keep.h>
+#include <kernel/unwind.h>
+
+.section .text
+/*
+ * int sm_pm_cpu_suspend(uint32_t arg, int (*fn)(uint32_t))
+ * @arg will be passed to fn as argument
+ * return value: 0 - cpu resumed from suspended state.
+ *		 -1 - cpu not suspended.
+ */
+FUNC sm_pm_cpu_suspend, :
+UNWIND( .fnstart)
+UNWIND( .cantunwind)
+	push	{r4 - r12, lr}
+	mov	r5, sp
+	sub	sp, sp, #SM_PM_CTX_SIZE
+	push	{r0, r1}
+
+	mov	r2, r5
+	mov	r1, #SM_PM_CTX_SIZE
+	add	r0, sp, #8
+	blx	sm_pm_cpu_suspend_save
+	adr	lr, aborted
+	/* Jump to arch specific suspend */
+	pop	{r0, pc}
+aborted:
+	/* cpu not suspended */
+	add	sp, sp, #SM_PM_CTX_SIZE
+	/* Return -1 to the caller */
+	mov	r0, #(-1)
+	pop	{r4 - r12, pc}
+UNWIND( .fnend)
+END_FUNC sm_pm_cpu_suspend
+
+LOCAL_FUNC sm_pm_cpu_v7_suspend, :
+UNWIND(.fnstart)
+UNWIND( .cantunwind)
+	push	{r4 - r11}
+	read_fcseidr r4
+	read_tpidruro r5
+	stmia	r0!, {r4 - r5}
+	read_dacr r4
+#ifdef CFG_WITH_LPAE
+#error "Not supported"
+#else
+	read_ttbr0 r5
+	read_ttbr1 r6
+	read_ttbcr r7
+#endif
+	read_sctlr r8
+	read_actlr r9
+	read_cpacr r10
+	read_mvbar r11
+	stmia	r0!, {r4 - r11}
+	read_prrr r4
+	read_nmrr r5
+	read_vbar r6
+	read_nsacr r7
+	stmia	r0, {r4 - r7}
+	pop	{r4 - r11}
+	bx	lr
+UNWIND( .fnend)
+END_FUNC sm_pm_cpu_v7_suspend
+
+FUNC sm_pm_cpu_do_suspend, :
+UNWIND(.fnstart)
+UNWIND(.cantunwind)
+	push	{r4, r5}
+	read_midr r4
+	ubfx	r5, r4, #4, #12
+	ldr     r4, =CORTEX_A7_PART_NUM
+	cmp	r5, r4
+	beq	a7_suspend
+	ldr     r4, =CORTEX_A9_PART_NUM
+	cmp	r5, r4
+	beq	a9_suspend
+	/* cpu not supported */
+	b	.
+	/* A9 needs PCR/DIAG */
+a9_suspend:
+	read_pcr r4
+	read_diag r5
+	stmia	r0!, {r4 - r5}
+a7_suspend:
+	pop	{r4, r5}
+	b	sm_pm_cpu_v7_suspend
+UNWIND( .fnend)
+END_FUNC sm_pm_cpu_do_suspend
+
+FUNC sm_pm_cpu_resume, :
+UNWIND(.fnstart)
+UNWIND( .cantunwind)
+	cpsid	aif
+
+	/* Call into get_core_pos */
+	adr	r0, _core_pos
+	ldr	r1, [r0]
+	add	r0, r0, r1
+	blx	r0
+
+	/*
+	 * At this point, MMU is not enabled now.
+	 * 1. Get the runtime physical address of _suspend_sp
+	 * 2. Get the offset from _suspend_sp to &thread_core_local
+	 * 3. Get the runtime physical address of thread_core_local
+	 * Since moving towards non-linear mapping,
+	 * `ldr r0, =thread_core_local` is not used here.
+	 */
+	adr	r4, _suspend_sp
+	ldr	r5, [r4]
+	add	r4, r4, r5
+
+	mov_imm r1, THREAD_CORE_LOCAL_SIZE
+	mla	r0, r0, r1, r4
+
+	ldr	r0, [r0, #THREAD_CORE_LOCAL_SM_PM_CTX_PHYS]
+	/* Need to use r0!, because sm_pm_cpu_do_resume needs it */
+	ldmia	r0!, {sp, pc}
+UNWIND( .fnend)
+END_FUNC sm_pm_cpu_resume
+
+FUNC sm_pm_cpu_resume_mmu, :
+UNWIND(.fnstart)
+UNWIND( .cantunwind)
+	isb
```

Comment:
The `isb` is either needed or not needed, in this case it's not needed. It would be clearer if this code was at the end of `cpu_do_resume()`. Please fold `cpu_resume_mmu()` into the end of `cpu_do_resume()`.

What I'm asking for is better symmetry between the suspend and resume functions, If you'd rather add a `cpu_v7_resume()` than folding `cpu_v7_suspend()` into `cpu_do_suspend()` that's fine.

Reply:
ok. I'll move cpu_v7_susend/resume into cpu_do_suspend/resume.